app/grandchallenge/cases/image_builders/tiff.py

```diff
@@ -214,52 +250,196 @@ def _add_folder_uploads(*, dzi_output, image, new_folder_upload):
+# ImageFile =, ImageFile(x,y) ImageFile(z,x,y)
+vms_pattern = r"ImageFile(\(\d*,\d*(,\d)?\))?\s?=\s?.|MapFile\s?=\s?.|OptimisationFile\s?=\s?.|MacroImage\s?=\s?."
+
+
+def _compile_mrx(files: List[Path], converter) -> List[GrandChallengeTiffFile]:
+    def get_filenames_from_ini(ini_file: Path):
+        file_matched = []
+        if not ini_file.exists():
+            ini_file = [
+                f
+                for f in ini_file.parent.iterdir()
+                if f.name.lower() == ini_file.name.lower()
+            ][0]
+        with open(ini_file, "r") as f:
+            lines = [
+                line for line in f.readlines() if re.match(mirax_pattern, line)
+            ]
+            for line in lines:
+                original_name = line.split("=")[1].strip()
+                file_matched.append(ini_file.parent / original_name)
+            file_matched.append(ini_file)
+            return file_matched
+
+    # OpenSlide expects the following for a Mirax file:
+    # The filename ends with .mrxs.
+    # A directory exists in the same location as the file, with the same name as the
+    # file minus the extension.
+    # A file named Slidedat.ini exists in the directory.
+    compiled_files: List = []
+    for file in files:
+        try:
+            gc_file = GrandChallengeTiffFile(file)
+
+            # Find Slidedat.ini, which provides us with all the other file names
+            name, _ = os.path.splitext(file.name)
+            slide_dat = file.parent / name / "Slidedat.ini"
+            matching_files = get_filenames_from_ini(slide_dat)
+            # convert to tif
+            tiff_file = _convert_to_tiff(
+                path=file, pk=gc_file.pk, converter=converter
+            )
+        except Exception:
+            continue
+        else:
+            gc_file.path = tiff_file
+            gc_file.associated_files = matching_files
+            gc_file.associated_files.append(file)
+            compiled_files.append(gc_file)
+
+    return compiled_files
+
+
+def _compile_vms(files: List[Path], converter) -> List[GrandChallengeTiffFile]:
+    def get_filenames_from_vms(vms_file: Path):
+        file_matched = []
+        with open(str(vms_file.absolute()), "r") as f:
+            lines = [
+                line for line in f.readlines() if re.match(vms_pattern, line)
+            ]
+            for line in lines:
+                original_name = line.split("=")[1].strip()
+                if os.path.exists(vms_file.parent / original_name):
+                    file_matched.append(vms_file.parent / original_name)
+            return file_matched
+
+    compiled_files: List = []
+    for file in files:
+        try:
+            gc_file = GrandChallengeTiffFile(file)
+            vms_files = get_filenames_from_vms(gc_file.path)
+            tiff_file = _convert_to_tiff(
+                path=file, pk=gc_file.pk, converter=converter
+            )
+        except Exception:
+            continue
+        else:
```

Comment:
You could possibly create a separate function for this, as you repeat this bit of code 3 times (with some very minor variations). This is of course a bit nitpickey.

Reply:
good point